README.rst
===============================
certbot-dns-transip
===============================

.. image:: https://readthedocs.org/projects/certbot-dns-transip/badge/?version=stable
   :target: https://certbot-dns-transip.readthedocs.io/en/stable/?badge=stable
   :alt: Documentation Status
   
.. image:: https://www.travis-ci.org/hsmade/certbot-dns-transip.svg?branch=master&status=passed
   :target: https://www.travis-ci.org/github/hsmade/certbot-dns-transip
   :alt: Build Status

Certbot plugin to authenticate using dns TXT records via Transip API


* Documentation: https://readthedocs.org/projects/certbot-dns-transip/

You can also run this directly from Docker, and get the certificates and keys written to disk for further processing.

For example the following command can be used. This assumes the `transip.ini` file and the keyfile are present in `/tmp/letsencrypt`. ::

    docker run -ti -v `/tmp/letsencrypt`:/etc/letsencrypt \
        hsmade/certbot-transip \
        certonly -n \
        -d 'your.domain.com' \
        -a certbot-dns-transip:dns-transip \
        --certbot-dns-transip:dns-transip-credentials /etc/letsencrypt/transip.ini \
        --certbot-dns-transip:dns-transip-propagation-seconds 240 \
        -m your@domain.com \
        --agree-tos \
        --eff-email


HISTORY.rst
.. :changelog:

History
-------

0.4.3 (13-10-2021)
---------------------

* add `dns_transip_global_key` as option in ini file


0.4.1 (07-10-2021)
---------------------

* refactored code to use new v6 API
* changed workflow to github actions
* updated docker base image to python:3.9

0.3.0 (24-03-2020)
---------------------

* updated CI and docker to python 3.6
* changed dockerfile from using CMD to ENTRYPOINT


0.2.6 (21-03-2020)
---------------------

* updated transip lib

0.1.4 (18-01-2019)
---------------------

* Added docker image + instructions

0.1.2 (12-11-2018)
---------------------

* Updated the API for transip to current version

0.1 (14-07-2017)
---------------------

* First release on https://pypi.python.org/pypi/certbot-dns-transip


tests/test_certbot_dns_transip.py
from unittest import TestCase

import certbot
from certbot.plugins import dns_test_common
from certbot.plugins.dns_test_common import DOMAIN
from certbot.tests import util as test_util
from certbot_dns_transip.dns_transip import _TransipClient
import mock
import os
from tempfile import mktemp
from certbot.errors import PluginError
import logging

logging.getLogger(_TransipClient.__class__.__name__).setLevel('DEBUG')

USERNAME = 'foobar'
KEY_FILE = mktemp()


# wrap the class we want to test, to remove the client init in __init__ (as it will break)
class _TransipClientTest(_TransipClient):
    def __init__(self):
        self.logger = logging.getLogger(__name__)


class _DomainMock:
    def __init__(self, name):
        self.name = name


class Test_TransipClient(TestCase):
    def setUp(self):
        self.client = mock.MagicMock()
        with open(KEY_FILE, 'w') as key_file:
            key_file.write('''-----BEGIN RSA PRIVATE KEY-----
foobar
-----END RSA PRIVATE KEY-----''')
        self.transip_client = _TransipClientTest()
        self.transip_client.client = self.client
        self.transip_client.client.domains = self.client
        self.transip_client.client.domains.list.return_value = [_DomainMock(name="example.com")]
        self.add_record = {"name": "test.test", "type": "TXT", "content": "new record", "expire": 1}

    def tearDown(self):
        os.unlink(KEY_FILE)

    def test_add_txt_record(self):
        domain = mock.MagicMock()

        self.client.domains.get.return_value = domain
        self.transip_client.add_txt_record(
            domain_name='example.com',
            record_content='new record',
            record_name='test.test.example.com'
        )
        domain.dns.create.assert_called_once_with(self.add_record)

    def test_del_txt_record(self):
        domain = mock.MagicMock()

        self.client.domains.get.return_value = domain
        self.transip_client.del_txt_record(
            domain_name='example.com',
            record_content='new record',
            record_name='test.test.example.com'
        )
        domain.dns.delete.assert_called_once_with(self.add_record)

    def test__find_domain(self):
        self.assertEquals(self.transip_client._find_domain('example.com'), 'example.com')

    def test__find_domain_fail(self):
        self.assertRaises(PluginError, self.transip_client._find_domain, 'example2.com')

    def test__compute_record_name(self):
        self.assertEquals(_TransipClient._compute_record_name('example.com', 'record.sub.example.com'), 'record.sub')


class AuthenticatorTest(test_util.TempDirTestCase, dns_test_common.BaseAuthenticatorTest):
    def setUp(self):
        from certbot_dns_transip.dns_transip import Authenticator

        super(AuthenticatorTest, self).setUp()

        path = os.path.join(self.tempdir, 'file.ini')
        dns_test_common.write({"transip_key_file": KEY_FILE, "transip_username": USERNAME}, path)

        self.config = mock.MagicMock(transip_credentials=path,
                                     transip_propagation_seconds=0)  # don't wait during tests

        self.auth = Authenticator(self.config, "transip")

        self.mock_client = mock.MagicMock()
        # _get_transip_client | pylint: disable=protected-access
        self.auth._get_transip_client = mock.MagicMock(return_value=self.mock_client)

    def test_perform(self):
        certbot._internal.display.obj.get_display = mock.MagicMock()
        self.auth.perform([self.achall])

        expected = [mock.call.add_txt_record(DOMAIN, '_acme-challenge.' + DOMAIN, mock.ANY)]
        self.assertEqual(expected, self.mock_client.mock_calls)

    def test_cleanup(self):
        # _attempt_cleanup | pylint: disable=protected-access
        self.auth._attempt_cleanup = True
        self.auth.cleanup([self.achall])

        expected = [mock.call.del_txt_record(DOMAIN, '_acme-challenge.' + DOMAIN, mock.ANY)]
        self.assertEqual(expected, self.mock_client.mock_calls)


certbot_dns_transip/dns_transip.py
# -*- coding: UTF-8 -*-
# File: dns_transip.py
"""certbot DNS plugin for Transip."""

import logging
import os
from tempfile import mktemp
from distutils.util import strtobool

import transip
import zope.interface
from certbot import errors
from certbot import interfaces
from certbot.plugins import dns_common

__author__ = '''Wim Fournier <wim@fournier.nl>'''
__docformat__ = 'plaintext'
__date__ = '''14-07-2017'''

LOGGER = logging.getLogger(__name__)
TRANSIP_EXCEPTIONS = (
    transip.exceptions.TransIPError,
    transip.exceptions.TransIPHTTPError,
    transip.exceptions.TransIPIOError,
    transip.exceptions.TransIPParsingError
)


@zope.interface.implementer(interfaces.IAuthenticator)
@zope.interface.provider(interfaces.IPluginFactory)
class Authenticator(dns_common.DNSAuthenticator):

    """
    DNS Authenticator for Transip.

    This Authenticator uses the Transip API to fulfill a dns-01 challenge.
    """

    description = 'Obtain certs using a DNS TXT record (if you are using Transip for DNS).'

    def __init__(self, *args, **kwargs):
        """Setup object."""
        super(Authenticator, self).__init__(*args, **kwargs)
        self.credentials = None
        self.logger = LOGGER.getChild(self.__class__.__name__)
        self.temp_file = None

    @classmethod
    def add_parser_arguments(cls, add, **_):  # pylint: disable=arguments-differ
        super(Authenticator, cls).add_parser_arguments(add, default_propagation_seconds=240)
        add('credentials', help='Transip credentials INI file.')

    def more_info(self):
        """Returns info about this plugin."""
        return 'This plugin configures a DNS TXT record to respond to a dns-01 challenge using ' + \
               'the Transip API.'

    def _setup_credentials(self):
        self.credentials = self._configure_credentials(
            'credentials',
            'Transip credentials INI file',
            {
                'key_file': 'RSA key file'
                            '(convert with openssl rsa -in transip.key -out decrypted_key)',
                'username': 'Transip username',
            }
        )

    def _perform(self, domain, validation_name, validation):
        self.logger.debug('_perform: running adding txt record %s.%s', domain, validation_name)
        self._get_transip_client().add_txt_record(domain, validation_name, validation)

    def _cleanup(self, domain, validation_name, validation):
        self.logger.debug('_cleanup: removing adding txt record %s.%s', domain, validation_name)
        try:
            self._get_transip_client().del_txt_record(domain, validation_name, validation)
        except Exception:
            if self.temp_file:
                os.unlink(self.temp_file)
            raise

    def _get_transip_client(self):
        username = self.credentials.conf('username')
        global_key = False
        try:
            global_key = bool(strtobool(self.credentials.conf('global_key')))
        except ValueError:
            raise ValueError("dns_transip_global_key should have either 'yes' or 'no' as value")
        except AttributeError:  # global_key was not present in the config, use default
            pass

        if not self.credentials.conf('key_file'):
            if self.credentials.conf('rsa_key'):
                key_file = mktemp()
                os.chmod(key_file, 600)
                with key_file as key:
                    key.write(self.credentials.conf('rsa_key'))
            else:
                raise ValueError('Please specify either an RSA key, or an RSA key file')
        else:
            key_file = self.credentials.conf('key_file')
        self.logger.debug('Creating Transip API client for user %s', username)
        return _TransipClient(username=username, key_file=key_file, global_key=global_key)


class _TransipClient:
    """Encapsulates all communication with the Transip API."""

    def __init__(self, username, key_file, global_key):
        self.logger = LOGGER.getChild(self.__class__.__name__)
        self.client = transip.TransIP(login=username, private_key_file=key_file, global_key=global_key)

    def add_txt_record(self, domain_name, record_name, record_content):
        """
        Add a TXT record using the supplied information.

        :param str domain_name: The domain to use to associate the record with.
        :param str record_name: The record name (typically beginning with '_acme-challenge.').
        :param str record_content: The record content (typically the challenge validation).
        :raises certbot.errors.PluginError: if an error occurs communicating with the Transip API
        """
        canonical_domain = self._find_domain(domain_name)

        try:
            domain = self.client.domains.get(canonical_domain)
        except TRANSIP_EXCEPTIONS as error:
            raise errors.PluginError('Error finding domain using the Transip API: {0}'.format(error))

        new_record = {
            "name": self._compute_record_name(canonical_domain, record_name),
            "type": "TXT",
            "content": record_content,
            "expire": 1,
        }

        try:
            domain.dns.create(new_record)
        except TRANSIP_EXCEPTIONS as error:
            raise errors.PluginError('Error finding domain using the Transip API: {0}'.format(error))

    def del_txt_record(self, domain_name, record_name, record_content):
        """
        Delete a TXT record using the supplied information.

        Note that both the record's name and content are used to ensure that similar records
        created concurrently (e.g., due to concurrent invocations of this plugin) are not deleted.

        Failures are logged, but not raised.

        :param str domain_name: The domain to use to associate the record with.
        :param str record_name: The record name (typically beginning with '_acme-challenge.').
        :param str record_content: The record content (typically the challenge validation).
        """
        canonical_domain = self._find_domain(domain_name)

        try:
            domain = self.client.domains.get(canonical_domain)
        except TRANSIP_EXCEPTIONS as error:
            raise errors.PluginError('Error finding domain using the Transip API: {0}'.format(error))

        delete_record = {
            "name": self._compute_record_name(canonical_domain, record_name),
            "type": "TXT",
            "content": record_content,
            "expire": 1,
        }

        try:
            domain.dns.delete(delete_record)
        except TRANSIP_EXCEPTIONS as error:
            raise errors.PluginError('Error finding domain using the Transip API: {0}'.format(error))

    def _find_domain(self, domain_name):
        """
        Find the domain object for a given domain name.

        :param str domain_name: The domain name for which to find the corresponding Domain.
        :returns: The Domain, if found.
        :rtype: `str`
        :raises certbot.errors.PluginError: if no matching Domain is found.
        """
        domain_name_guesses = dns_common.base_domain_name_guesses(domain_name)

        try:
            domains = [item.name for item in self.client.domains.list()]
        except TRANSIP_EXCEPTIONS as error:
            raise errors.PluginError('Error finding domain using the Transip API: {0}'.format(error))

        if not domains:
            raise errors.PluginError("Transip API returned no domains")

        for guess in domain_name_guesses:
            if guess in domains:
                self.logger.debug('Found base domain for %s using name %s', domain_name, guess)
                return guess

        raise errors.PluginError('Unable to determine base domain for {0} using names: {1} and domains: {2}.'
                                 .format(domain_name, domain_name_guesses, domains))

    @staticmethod
    def _compute_record_name(domain, full_record_name):
        # The domain, from Transip's point of view, is automatically appended.
        return full_record_name.rpartition("." + domain)[0]
